extensions/ql-vscode/src/view/common/telemetry.ts

```diff
@@ -0,0 +1,60 @@
+import { useEffect, useMemo, useRef } from "react";
+import { vscode } from "../vscode-api";
+
+/**
+ * A react effect that outputs telemetry events whenever the value changes.
+ *
+ * @param value Default value to pass to React.useState
+ * @param telemetryAction Name of the telemetry event to output
+ * @param options Extra optional arguments, including:
+ *   filterTelemetryOnValue: If provided, only output telemetry events when the
+ *       predicate returns true. If not provided always outputs telemetry.
+ *   debounceTimeout: If provided, will not output telemetry events for every change
+ *       but will wait until specified timeout happens with no new events ocurring.
+ */
+export function useTelemetryOnChange<S>(
+  value: S,
+  telemetryAction: string,
+  options?: {
+    filterTelemetryOnValue?: (value: S) => boolean;
+    debounceTimeoutMillis?: number;
+  },
+) {
+  const previousValue = useRef(value);
+  const filterTelemetryOnValue = options?.filterTelemetryOnValue;
+  const debounceTimeoutMillis = options?.debounceTimeoutMillis;
```

Comment:
I think this can be simplified slightly:

```suggestion
  {
    filterTelemetryOnValue,
    debounceTimeoutMillis,
  }: {
    filterTelemetryOnValue?: (value: S) => boolean;
    debounceTimeoutMillis?: number;
  } = {},
) {
  const previousValue = useRef(value);
```

That should also allow you to use `@param` on the names of the properties, rather than on `options`.

Reply:
Nice. That is cleaner.